Coordinate maps take function of time as unique ptr

nilsdeppe opened the pull request:
## Proposed changes

- Previously the coordinate maps took functions of time by reference to the base class. This changes that to `unique_ptr` since that is what we will actually have in the DataBox.

Depends on:
- #1845 in the sense that there are some include conflicts. Please review only the unique_ptr commit.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Factor ProductMaps.hpp into a tpp file

nilsdeppe committed:
Switch function of time coord map args to unique_ptr

Previously the argument type of an unordered_map<string, FunctionOfTime&>.
However, in practice we will have an unordered_map<string,
unique_ptr<FunctionOfTime>> in the DataBox that gets updated. Rather than having
some intermediate compute item compute item that creates a new unordered_map.

moxcodes commented on src/Domain/CoordinateMaps/CoordinateMap.hpp:
```diff
@@ -98,10 +98,12 @@ class CoordinateMapBase : public PUP::able {
+      const std::unordered_map<
+          std::string,
+          std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
           functions_of_time = std::unordered_map<
```
is this default required? It seems as though it could potentially lead to confusing behavior. My understanding is that if the derived class is passed as a `CoordinateMap*` and these functions are called on it without the `functions_of_time`, it would then pass the defaulted (empty) map to the derived class implementation, which may or may not try to access the contents depending on how the derived class is supposed to behave.

nilsdeppe replied:
That's right. We want the classes to be able to wrap both time-dependent and time-independent maps. The caller of these maps is responsible for knowing whether the maps are time-dependent. Since this will generically be handled by compute items and nobody should call the maps directly I don't see it causing problems in practice.

moxcodes replied:
I see. I suppose if the interface was widely used, I might suggest a solution with more direct overloads function overloads rather than defaults, but it's probably not necessary if the compute items are handling it; regardless I think that sort of change is out of scope for this PR.

nilsdeppe replied:
Sorry, how would an overload by any different from defaults?

moxcodes replied:
My thinking was that an overloaded pure virtual would force the derived classes to more explicitly handle the case they're not suited to; the time-dependent ones would have to make choices about what to do when they don't have a `functions_of_time` map (probably just error with a tailored message), so it would be harder to overlook than the current virtuals where I could see a derived class be implemented that just grabs things from the map failing at runtime with more challenging errors. I think It's only a minor improvement, though, so without a better idea, it's probably not worth the time.

nilsdeppe replied:
I think this is a good discussion to have because right now (even in a follow up PR) the interface is still reasonable to change. Let me go over in detail what currently happens:

The `CoordinateMapBase` class only ever has `CoordinateMap<Maps...>` as the derived classes. The `CoordinateMap` class calls the individual coordinate maps, checking whether they require a time argument or not. We could have an `ASSERT` in the functions that calls the coordinate maps that if they are time-dependent that `t` is not the default value (signaling nan) and that the map is not empty. Now, I think checking for a nan with a signalling nan causes an FPE, but that's easily worked around. Hopefully this makes it clear how the maps are working.

In terms of overloading, are you suggesting:
```
operator()(coords);

operator()(coords, time, functions_of_time);
```
? I'm not sure how to make this any different from the current implementation since by the very nature of the base class we erase the knowledge of whether or not the derived maps have any time dependence. That means we can only check at runtime.

moxcodes replied:
hm, okay, that brings up other questions.
- It appears that, for instance, `CubicScale` has the same interface that is outlined in `CoordinateMapBase`, but it doesn't inherit, so it must never be used in the generic way that I'm thinking of, but it's unclear to me why. In my suggestion, I was thinking of all specific maps needing to inherit from the `CoordinateMapBase`, but that doesn't seem to be the case.
- It also appears that there are a bunch of overloads (or make_overloads) already in impls for `CoordinateMap...`. If there's already a metafunction for determining time dependence, I think you can use the impl structure to make this a compile-time check with a like the following:
```
TensorType operator()(TensorType coords) noexcept {
  return paren_impl(coords,
                    CoordinateMap_detail::is_map_time_dependent_t<Maps>{});
}
template <typename IsTimeDependent>
TensorType paren_impl(TensorType coords, IsTimeDependent /*meta*/) {
  static_assert(not IsTimeDependent::value,
                "The called map is time dependent, so a function of time must "
                "be supplied");
  /// the rest of the function...
}
```
The reverse could probably be done to prevent a time-independent map from being called with a function of time supplied, but I don't actually know if that version is desirable.

edit:formatted the code block so it's easier to read.

nilsdeppe replied:
- Right, so all coordinate maps will only be used through the `CoordinateMap` class, which inherits off of `CoordinateMapBase`. The motivation was so that rather than having to deal with supporting double/DataVector implementations of virtual functions in all coordinate maps, it's only needed in the "master" versions (`CoordinateMap` and its base). Furthermore, it means fewer places to deal with serializing base/derived classes.
- Yes, so `CoordinateMap` checks each map it composes to see if it should call it with or without the time arguments. We ultimately want to be able to compose both time independent and time-dependent maps together in any order. This means that unless we want to template `Domain` on the individual maps, and thus also the parallel components, we need the maps to be runtime adjustable. What this means is that any such `static_assert` would be triggered all the time since we don't know at compile time whether the maps will be called with or without the arguments. It might be possible to enable such a check without templating everything on the maps, but I would need to think quite a bit more about that to be sure :) The tricky thing is that since the virtual functions can't be templates they will always get instantiated, whether or not they're called, so sticking a static assert in them can easily become `static_assert(false, "Now we never compile");` (I've encountered this and was sad :P )

moxcodes replied:
I see. Given that this inheritance is so limited, I'm far less concerned about confusing behavior.

Now, I can see how the set of maps that are called from a particular site might change at runtime; presumably by shuffling around the base class pointers. I also agree that the static assert definitely can't appear in the overridden virtual. But,is it not the case that any case `CoordinateMap_detail::is_map_time_dependent_t<Map>` is `true_type` that the corresponding map functions will runtime error with `unordered_map` access error when called with an empty `unordered_map`? If not, then I've misunderstood the metafunction, but if so it seems like it offers at least a little extra safety to demand at compile time that at least something is passed in for a function of time when it's known that a map that gets called is going to need it.

nilsdeppe replied:
Yes, I think adding `ASSERT`s to check for non-empty functions of time and non-NaN time is a good idea. I've added #1862 and assigned it to me. Once some of the other changes to `CoordinateMap` are in I'll add these checks :)

moxcodes replied:
I'm happy with that solution, thanks!

nilsdeppe replied:
You're welcome! Thank you for suggesting it :)

fmahebert commented on tests/Unit/Domain/FunctionsOfTime/Test_SettleToConstant.cpp:
```diff
@@ -13,7 +13,7 @@
 
 namespace {
-void test(const domain::FunctionsOfTime::FunctionOfTime& f_of_t,
+void test(const std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime> f_of_t,
```
Should this argument be passed by reference? I worry you are copying an owning unique ptr.

nilsdeppe replied:
This is done after bringing in #1856 :+1:

moxcodes commented on tests/Unit/Domain/CoordinateMaps/Test_CoordinateMap.cpp:
```diff
@@ -902,14 +902,13 @@ void test_time_dependent_map() {
+  std::unordered_map<std::string,
+                     std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
+      functions_of_time{};
+  functions_of_time["trans"] =
```
Does the initializer list not work for this one? it would be preferable to const with initializer list if it works. (same question for a few other tests below)

nilsdeppe replied:
It does not, or I could not convince both clang and GCC to agree on the syntax. There's some internal copy going on in the implementation of unordered_map (I'm not sure whether that's a bug in the STL or not)

wthrowe replied:
Initializer lists are always const, so they're questionable-to-broken with anything with a non-trivial (or missing) copy operation.

My understanding is that this is because they were developed concurrently with move semantics so didn't consider them in the design.  I've never heard a reason why the standards committee hasn't fixed this in subsequent versions.

fmahebert commented on tests/Unit/Domain/CoordinateMaps/Test_Translation.cpp:
```diff
@@ -31,13 +31,16 @@ SPECTRE_TEST_CASE("Unit.Domain.CoordMapsTimeDependent.Translation",
+  using Polynomial = domain::FunctionsOfTime::PiecewisePolynomial<deriv_order>;
+  using FoftPtr = std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>;
+  std::unordered_map<std::string,
+                     std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
```
Why not use `FoftPtr` here?

fmahebert replied:
@nilsdeppe I think you missed this one?

nilsdeppe replied:
Self-rekt :P Thanks :)

fmahebert commented on tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp:
```diff
@@ -208,26 +206,25 @@ void test_boundaries() {
+        domain::FunctionsOfTime::PiecewisePolynomial<deriv_order>;
+    using FoftPtr = std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>;
+    std::unordered_map<std::string,
+                       std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
```
Why not use `FoftPtr` here?

fmahebert commented on tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp:
```diff
@@ -72,18 +71,17 @@ void test_map() {
+        domain::FunctionsOfTime::PiecewisePolynomial<deriv_order>;
+    using FoftPtr = std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>;
+    std::unordered_map<std::string,
+                       std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
```
Why not use `FoftPtr` here?

moxcodes commented on tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp:
```diff
@@ -72,18 +71,17 @@ void test_map() {
-    domain::FunctionsOfTime::FunctionOfTime& expansion_a_base = expansion_a;
+    using Polynomial =
+        domain::FunctionsOfTime::PiecewisePolynomial<deriv_order>;
+    using FoftPtr = std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>;
```
If you're going to make this type alias, why not put it in a wider scope so that it doesn't need to be redeclared multiple times? Also, if it's around, you may as well put it in the `unordered_map` declarations.
Not sure if it's readable enough to be worth bringing all the way up to the src, but it would shorten declarations there substantially.

nilsdeppe replied:
I personally like scoping things, in this case inside the function. I do agree that it should be used in the `unordered_map`s :)

fmahebert commented on tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp:
```diff
@@ -31,28 +31,27 @@ void cubic_scale_non_invertible(const double a0, const double b0,
+  using Polynomial = domain::FunctionsOfTime::PiecewisePolynomial<deriv_order>;
+  using FoftPtr = std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>;
+  std::unordered_map<std::string,
+                     std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
```
Why not use `FoftPtr` here?

moxcodes commented on tests/Unit/Domain/FunctionsOfTime/Test_SettleToConstant.cpp:
```diff
@@ -13,7 +13,7 @@
 
 namespace {
-void test(const domain::FunctionsOfTime::FunctionOfTime& f_of_t,
+void test(const std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime> f_of_t,
```
wait... is `f_of_t` used in this test function? How does this work without replacing a bunch of `.`s with `->`? (.. and why do the tests pass without that change?)

nilsdeppe replied:
I can bring #1856 into this PR. probably would be better.

moxcodes replied:
Ah, I see now that it's corrected in that commit. Yeah, I think it would be better to get the tests all sorted in a single PR.

moxcodes reviewed (changes requested):
Most of these are weak suggestions, I'm willing to approve without most of the suggested alterations. The exception is the parts in `Test_SettleToConstant.cpp` where the `unique_ptr` appears to be accessed incorrectly.

fmahebert commented on src/Domain/CoordinateMaps/CoordinateMap.hpp:
```diff
@@ -74,18 +70,22 @@ class CoordinateMapBase : public PUP::able {
-                               domain::FunctionsOfTime::FunctionOfTime&>&
+      const std::unordered_map<
+          std::string,
+          std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
```
Here it looks like you're replacing a map of non-owning references with a map of owning pointers. Can you comment (just here in Github, to help me understand) where the change in the call site is? I see the corresponding change in the test, but nowhere else. Is this functionality simply not used in the source code yet?

nilsdeppe replied:
Yes, so currently we don't have any moving  mesh evolutions, only some tests for the maps. What this PR is addressing is that when we add the functions of time into the DataBox we will add them as `unique_ptr`s through the base class, which means a more natural interface for these functions is to take the `unique_ptr`s rather than references. I'm currently working on getting the time dependence into the evolution executables.

fmahebert commented:
Oh, looks like I mostly just repeated @moxcodes' review 😛

nilsdeppe commented:
Thanks for the reviews! 

I brought in the fixes from #1856 which I squashed directly since I think it makes more sense to review the couple files for the F of Ts again given that before they were broken (sorry...). I added a fixup for the `FoftPtr`

fmahebert reviewed (commented):
I am happy for you to squash (including this)

moxcodes commented:
I'm happy for you to squash the fixup.

nilsdeppe commented:
Okay, squashed :) Thanks for the super quick and thorough reviews!! :D

wthrowe reviewed (commented):
Not a review.  Just answering a tangential question.

nilsdeppe commented:
I missed removing a closing `>` in `Test_Translation.cpp` when switching to `FoftPtr`... :(
